bigchaindb/models.py

```diff
@@ -0,0 +1,208 @@
+                    raise DoubleSpend('input `{}` was already spent'
+                                      .format(input_txid))
+
+                input_conditions.append(input_tx.conditions[input_cid])
+        else:
+            allowed_operations = ', '.join(Transaction.ALLOWED_OPERATIONS)
+            raise TypeError('`operation`: `{}` must be either {}.'
+                            .format(self.operation, allowed_operations))
+
+        if not self.fulfillments_valid(input_conditions):
+            raise InvalidSignature()
+        else:
+            return self
+
+
+class Block(object):
+    def __init__(self, transactions=None, node_pubkey=None, timestamp=None,
+                 voters=None, signature=None):
+        if transactions is not None and not isinstance(transactions, list):
+            raise TypeError('`transactions` must be a list instance or None')
+        else:
+            self.transactions = transactions or []
+
+        if voters is not None and not isinstance(voters, list):
+            raise TypeError('`voters` must be a list instance or None')
+        else:
+            self.voters = voters or []
+
+        if timestamp is not None:
+            self.timestamp = timestamp
+        else:
+            self.timestamp = gen_timestamp()
+
+        self.node_pubkey = node_pubkey
+        self.signature = signature
+
+    def __eq__(self, other):
+        try:
+            other = other.to_dict()
+        except AttributeError:
+            return False
+        return self.to_dict() == other
+
+    def validate(self, bigchain):
+        """Validate a block.
+
+        Args:
+            bigchain (Bigchain): an instantiated bigchaindb.Bigchain object.
+
+        Returns:
+            block (Block): The block as a `Block` object if it is valid.
+                           Else it raises an appropriate exception describing
+                           the reason of invalidity.
+
+        Raises:
+            OperationError: if a non-federation node signed the block.
+        """
+
+        # First, make sure this node hasn't already voted on this block
+        if bigchain.has_previous_vote(self.id, self.voters):
+            return self
+
+        # Check if the block was created by a federation node
+        possible_voters = (bigchain.nodes_except_me + [bigchain.me])
+        if self.node_pubkey not in possible_voters:
+            raise OperationError('Only federation nodes can create blocks')
+
+        if not self.is_signature_valid():
+            raise InvalidSignature('Block signature invalid')
+
+        # Finally: Tentative assumption that every blockchain will want to
+        # validate all transactions in each block
+        for tx in self.transactions:
+            # NOTE: If a transaction is not valid, `is_valid` will throw an
+            #       an exception and block validation will be canceled.
+            bigchain.validate_transaction(tx)
+
+        return self
+
+    def sign(self, private_key):
```

Comment:
@diminator a side note: it would be useful if we had some kind of reference(s) for the usage of these names. I did see a bit that the terms "signing key" and "verification (or verifying) key" tend to be used in some literature about cryptology.